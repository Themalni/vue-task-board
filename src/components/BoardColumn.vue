<template>
  <div class="column">
    <div class="column-header">{{ category.title }}</div>
    <div class="column-body">
      <draggable v-model="items" group="task-items" class="draggable-container">
        <Card
          v-for="task in items"
          :key="task.id"
          :description="task.description"
          @openCard="openTask(task)"
        >
          <Tag class="mt-2" :type="task.type"/>
        </Card>
      </draggable>
      <View-Task-Modal v-if="viewModalIsOpen"></View-Task-Modal>
    </div>
  </div>
</template>

<script>
import draggable from 'vuedraggable'
import Card from '@/components/Card'
import Tag from '@/elements/Tag'
import ViewTaskModal from '@/components/task/ViewTaskModal'
import { mapGetters, mapActions } from 'vuex'

export default {
  name: 'BoardColumn',
  components: {
    draggable,
    Card,
    Tag,
    ViewTaskModal
  },
  props: {
    category: {
      type: Object
    }
  },
  computed: {
    ...mapGetters(['viewModalIsOpen', 'activeTask']),
    items: {
      get () {
        return this.category.tasks
      },
      set (value) {
        const payload = {
          id: this.category.id,
          title: this.category.title,
          tasks: value
        }
        this.$store.dispatch('moveTask', payload)
      }
    }
  },
  methods: {
    ...mapActions(['changeViewTaskModalState', 'setActiveCard', 'moveTask']),
    openTask (task) {
      this.$store.dispatch('changeViewTaskModalState', true)
      this.$store.dispatch('setActiveCard', task)
    }
  }
}
</script>

<style lang="scss">
.column {
  display: flex;
  flex-direction: column;
  width: 250px;
  min-height: 100%;
  margin-right: 0.5em;
  padding: 1em 0.5em 0.7em;
  border-radius: 4px;
  background-color: $grey;
  box-shadow: 0 1px 5px -3px $graphite;
}
.column-header {
  font-size: 1.1em;
  padding-bottom: 1em;
}
.column-body {
  min-height: 110px;
}
.draggable-container {
  min-height: 100%;
}
</style>
