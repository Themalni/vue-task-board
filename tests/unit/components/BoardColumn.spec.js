import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import draggable from 'vuedraggable'
import mockState from '../utils/mockState'
import BoardColumn from '@/components/BoardColumn'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(draggable)

const categoryTasks = [
  {
    id: '2e037380-d748-12e6-99be-5b4b2b69f7e9',
    categoryId: 'c1',
    description: 'Opis zadania #1',
    type: 'feature',
    status: 'Oczekujące'
  },
  {
    id: '2e037380-d748-12e6-99be-5b4b2b69f7e9',
    categoryId: 'c1',
    description: 'Opis zadania #1',
    type: 'feature',
    status: 'Oczekujące'
  }
]

describe('BoardColumn', () => {
  let wrapper
  let store
  let actions
  let state
  let getters

  beforeEach(() => {
    state = mockState.state
    getters = {
      viewModalIsOpen: () => false,
      activeTask: () => {}
    }
    actions = {
      changeViewTaskModalState: jest.fn(),
      setActiveCard: jest.fn(),
      moveTask: jest.fn()
    }
    store = new Vuex.Store({
      state,
      getters,
      actions
    })

    wrapper = mount(BoardColumn, {
      propsData: {
        category: state.categories[0]
      },
      store,
      localVue,
      sync: false,
      stabs: { draggable }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('does contain `column` element', () => {
    expect(wrapper.classes('column')).toBe(true)
  })
  it('calls `openTask` when card is clicked', () => {
    wrapper.vm.openTask()
    expect(actions.changeViewTaskModalState.mock.calls).toHaveLength(1)
    expect(actions.setActiveCard.mock.calls).toHaveLength(1)
  })
})
