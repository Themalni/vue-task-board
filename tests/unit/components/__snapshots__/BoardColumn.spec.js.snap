// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardColumn renders correctly 1`] = `
<div class="column">
  <div class="column-header">OczekujÄ…ce</div>
  <div class="column-body">
    <div class="draggable-container">
      <div class="card">
        <div class="card-body">Opis zadania #1</div>
        <div class="card-footer">
          <div class="mt-2 tag text-sm tag-feature">
            feature
          </div>
        </div>
      </div>
      <div class="card">
        <div class="card-body">Opis zadania #2</div>
        <div class="card-footer">
          <div class="mt-2 tag text-sm tag-bugfix">
            bugfix
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
